plugins/lsp/schemastore.nix

```diff
@@ -0,0 +1,140 @@
+
+    extraOptions = let
+      schemaOpts = let
+        schemaEntry = types.submodule {
+          options = {
+            name = mkOption {
+              type = types.str;
+              description = "The name of the schema.";
+            };
+
+            description = helpers.mkNullOrStr "A description for this schema.";
+
+            fileMatch = helpers.mkNullOrOption (with types; either str (listOf str)) ''
+              Which filename to match against for this schema.
+            '';
+
+            url = mkOption {
+              type = types.str;
+              description = "The URL of this schema.";
+            };
+          };
+        };
+      in {
+        select = helpers.defaultNullOpts.mkListOf types.str "[]" ''
+          A list of strings representing the names of schemas to select.
+          If this option is not present, all schemas are returned.
+          If it is present, only the selected schemas are returned.
+          `select` and `ignore` are mutually exclusive.
+        '';
+
+        ignore = helpers.defaultNullOpts.mkListOf types.str "[]" ''
+          A list of strings representing the names of schemas to ignore.
+          `select` and `ignore` are mutually exclusive.
+        '';
+
+        replace = helpers.defaultNullOpts.mkAttrsOf schemaEntry "{}" ''
+          An attrs of elements representing schemas to replace with a custom schema.
+          The string key is the name of the schema to replace, the table value is the schema definition.
+          If a schema with the given name isn't found, the custom schema will not be returned.
+        '';
+
+        extra = helpers.defaultNullOpts.mkListOf schemaEntry "[]" ''
+          Additional schemas to include.
+        '';
+      };
+    in {
+      json = {
+        enable = mkOption {
+          type = types.bool;
+          default = true;
+          example = false;
+          description = "Whether to enable the json schemas in jsonls.";
+        };
+
+        settings = helpers.mkSettingsOption {
+          options = schemaOpts;
+          description = "Options supplied to the `require('schemastore').json.schemas` function.";
+          example = {
+            replace."package.json" = {
+              description = "package.json overridden";
+              fileMatch = ["package.json"];
+              name = "package.json";
+              url = "https://example.com/package.json";
+            };
+            extra = [
+              {
+                description = "My custom JSON schema";
+                fileMatch = "foo.json";
+                name = "foo.json";
+                url = "https://example.com/schema/foo.json";
+              }
+              {
+                description = "My other custom JSON schema";
+                fileMatch = ["bar.json" ".baar.json"];
+                name = "bar.json";
+                url = "https://example.com/schema/bar.json";
+              }
+            ];
```

Comment:
Seems odd to have `replace` and `extra` examples without `select` and `ignore`. Maybe they could be `yaml` examples? :thinking:

Reply:
I have not found existing examples online. I don't know if we should seek exhaustive examples.